lib/router/src/lib.rs

```diff
@@ -72,30 +82,63 @@ where
 {
     recognize: Rec,
     make: Mk,
-    cache: Mutex<Cache<Rec::Target, LoadShed<Mk::Value>>>,
+    cache: Lock<Cache<Rec::Target, LoadShed<Mk::Value>>>,
 }
 
-enum State<Req, Svc>
+struct RouteRequest<Req, Rec, Mk>
 where
-    Svc: svc::Service<Req>,
-    Svc::Error: Into<error::Error>,
+    Rec: Recognize<Req>,
+    Mk: Make<Rec::Target>,
+    Mk::Value: svc::Service<Req>,
+    <Mk::Value as svc::Service<Req>>::Error: Into<error::Error>,
 {
-    Init(Option<Req>, Svc),
-    Called(Svc::Future),
+    request: Option<Req>,
+    target: Option<Rec::Target>,
+    make: Option<Mk>,
```

Comment:
TIOLI: this is a _bit_ awkward, maybe putting all the things that get `take`n in a `struct` so you can have one `Option` you take in one place would be nicer? Or wrapping this whole struct in an `Option` in the State` enum? Not a big deal.

Reply:
fwiw, this was my suggestion and i think it's fine for now...

Reply:
There is also the case where we don’t want to take the make if we don’t need to. If the route is already in the cache, we don’t end up using it that field.

Reply:
Okay, that's fair. Carry on.